Update GPU occupancy checking to utilize CUDA's occupancy calculator

MattConley opened the pull request:
This change will make supporting future architectures easier and removes dependence on hard-coded GPU data

-Replace references to the UnqueryableDeviceParams struct with calls to CUDA's built-in occupancy calculation functions
-Update calls to the occupancy checking functions with the new changes
-Changes should provide more long-term reliability and will remove the need to manually update hard-coded data values for new GPU architectures

@tfboyd  @zheng-xq

MattConley committed:
Update GPU occupancy checking to utilize CUDA's occupancy calculator functions

-Replace references to the UnqueryableDeviceParams struct with calls to CUDA's built-in occupancy calculation functions
-Update calls to the occupancy checking functions with the new changes
-Changes should provide more long-term reliability and will remove the need to manually update hardcoded data values for new GPU architectures

jlebar commented:
Thanks for the PR, this looks really good.

My only concern is: Do you know when the new CUDA driver function you're calling was added?

jlebar commented on tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:
```diff
@@ -467,33 +467,26 @@ void CUDAExecutor::VlogOccupancyInfo(const KernelBase &kernel,
+      CompareOccupancy(&blocks_per_sm, device_description, regs_per_thread,
+                       smem_per_block, thread_dims, cufunc);
+  if (suggested_threads != 0) {
+    VLOG(2) << "The cuda occupancy calculator reccommends using "
```
typo

MattConley replied:
Ah my bad, thanks!

jlebar commented on tensorflow/stream_executor/device_description.h:
```diff
@@ -375,17 +327,18 @@ void CalculateDimensionality(const DeviceDescription &device_description,
+                     const DeviceDescription& device_description,
+                     uint64 registers_per_thread,
+                     uint64 shared_memory_per_block,
+                     const ThreadDim& thread_dims, CUfunction func);
```
This file is and these functions are platform-independent.  But the implementations of them are platform-*de*pendent.  So if anyone calls CalculateOccupancy they're going to get the wrong answer (or a crash) if they're not using CUDA.  In fact one were to build without CUDA support, StreamExecutor won't even link, as I read this.

This needs to be done somehow in a platform-independent way.

MattConley committed:
Move CUDA-specific occupancy calculation into proper file

-Maintain functionality, just move CalculateOccupancy() and CompareOccupancy() methods from device_description to cuda_gpu_executor
-Remove CUDA requirement in general class device_description

MattConley committed:
Fixed transition typo

MattConley committed:
Recommended typo fix

jlebar commented:
lgtm, but let's see what the tests say.

jlebar commented:
Looks like the tests are clean enough other than the clang-format business.  If you can make that change we should be able to merge this.

Thank you again for the patch, this is a good change.

MattConley committed:
Fixed clang formatting

aaroey commented:
Thanks @MattConley for the clang fix, but it seems there are still some of left, would you mind to fix them (again)?
Thanks for the patient.

MattConley committed:
Fully fixed clang errors

MattConley commented:
Apologies, I had overlooked the placement of some references and dereferences; this should be all of the required formats within the changed code

tensorflowbutler commented:
Nagging Assignee @aaroey: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

MattConley commented:
@aaroey @jlebar Thanks again for your help; is this ready to be merged in?

jlebar commented:
FYI I'm trying to merge this, running into some build problems internally, still not sure exactly what is going on.

MattConley commented:
Thanks for the update; I'm not getting the issues on my end, please let me know if I can do anything to help.

jlebar commented:
I'm not sure why I didn't see this when I reviewed the patch earlier, but this is actually a functional change for XLA.

Previously we'd set the block size to `device_desc.threads_per_core_limit() / device_desc.blocks_per_core_limit()`.  This is the smallest block size such that we can fill a core with these blocks.

Now we set the block size to the *max* block size: `device_desc.threads_per_block_limit();`.  That's the opposite of the old code.

Was this change intentional on your part?  I don't think it was intentional on my part...

Is there a CUDA API that gives us information that lets us recover the old behavior?  (I'm not seeing it.) 
 Or do we have to go back to hardcoding this info?

jlebar commented:
@MattConley friendly ping.  I don't want to revert this if I don't have to, but I'm not actually sure how to fix this.

MattConley commented:
Apologies for the delayed response; this change in behavior was my mistake.  It looks like the desired functionality is available through the  `cuOccupancyMaxActiveBlocksPerMultiprocessor` driver function, which is accessible from CUDADriver's `GetMaxOccupiedBlocksPerCore` wrapper.

I'm looking at adding back the `blocks_per_core_limit` functionality to the device description, and then letting the cuda_gpu_executor fill in the field using the above call.  I'm running into a slight issue of passing in a valid CUfunction, but aside from that it looks like this is the easiest way to achieve desired behavior while avoiding hard-coding the values.

I should have a solution fairly quickly, and will submit a merge request as soon as it's ready.

jlebar commented:
Hey, friendly ping on this.  (I can also pick it up if you're busy; like, there's no obligation, it's not like this is your job.  :)

MattConley commented:
Just submitted PR #24944 with a potential fix.  (Again, sorry for the delay :)  The solution seems somewhat clunky, but does integrate the occupancy calculator without the need for hardcoded values; definitely open to suggestions.
